### ISSUE TITLE
depends on policykit-1-gnome, which is unmaintained

### ISSUE THREAD

**Issue by smcv** [linuxmint/cinnamon#10172]
The Debian maintainers of polkit noticed that the Debian package for Cinnamon 4.x depends on `policykit-1-gnome`, and the same seems to be true in Linux Mint's packaging of Cinnamon 5.x.

PolicyKit-gnome hasn't been used by GNOME since GNOME 3 (around 2010), and hasn't been used by GNOME Flashback since 2015, so it is no longer maintained by GNOME. The upstream project was [officially deprecated 6 years ago](https://gitlab.gnome.org/Archive/policykit-gnome/-/commit/6295fec06e3bf1f87ac97763be5b5158cde31c4d) and it has now been moved to a read-only archived state. This doesn't seem great for something that Cinnamon is using as a security-sensitive desktop component.

Please consider alternatives: either giving Cinnamon a built-in polkit agent similar to the one in GNOME Shell (which appears to have been removed from Cinnamon in 2013), or forking PolicyKit-gnome under the Cinnamon umbrella and taking responsibility for its maintenance and security, or using some other polkit agent.

**Comment by Fantu** [comment]
@mtwebster @clefebvre why built-in polkit agent was removed?

**Comment by eli-schwartz** [comment]
I suppose this must be a problem with xfce4-session too.

```
$ pacman -Sii polkit-gnome| grep "Required By"
Required By     : cinnamon  xfce4-session
```


My only thought on the topic is, it is depressing that gnome continually abandons reusable components in favor of reimplementing features in non-reusable ways inside of gnome-shell or other leaf applications.

**Comment by smcv** [comment]
> I suppose this must be a problem with xfce4-session too

The xfce4-session in Debian doesn't have that dependency, so this could either be a bug in the Debian packaging (not having a dependency that it should) or in the Arch packaging (having a dependency that it doesn't need). I don't use XFCE, so I don't know who is correct here. (I don't use Cinnamon either, but it's reasonably obvious from the packaging and code that Cinnamon does rely on the polkit agent formerly used in GNOME.)

> gnome-shell or other leaf applications

GNOME Shell and its equivalents like Cinnamon aren't really leaf applications - they're the desktop shell that presents UI for desktop-level and system-level things.

**Comment by eli-schwartz** [comment]
> GNOME Shell and its equivalents like Cinnamon aren't really leaf applications - they're the desktop shell that presents UI for desktop-level and system-level things.

I don't see how this technicality affects my statement.

I added "or other leaf applications" after the fact as a token reference to packages found in a software "start menu" listing, which *also* display the same problems gnome-shell itself does. Maybe I should have said "packages" or "and generally with leaf applications" to make my point clearer.

**Comment by ItzSwirlz** [comment]
I'm not sure why the polkit agent was removed but I don't see any valid reason for it than "this existed so we didn't want to maintain it"

When I find the commit I'll make a PR to revert it, and see if I need to backport anything from GNOME

**Comment by ItzSwirlz** [comment]
Actually, it isn't used *at all.* to the best of what I can see. I'll have to run a few tests to see if it actually calls anything but I don't think it does.
![Screenshot from 2021-06-24 20-50-36](https://user-images.githubusercontent.com/24401303/123352754-ff653700-d52d-11eb-8ce4-94c11f3d94ac.png)

Nothing related to polkit, policykit, or polkitagent to the best of my knowledge isn't used at all. It's just an empty dependency and can also be removed from the meson.build file.

**Comment by eli-schwartz** [comment]
Are we talking about polkit-gnome or the polkit-agent libraries? I'm pretty sure the current state of affairs is that cinnamon doesn't contain any polkit code, but has a package level dependency on an external program that provides the right dbus interface to pop up a dialog box whenever polkit requests occur.

That's all this ticket is about, not any potential and completely unrelated pkg-config dependencies that are being searched for and linked into cinnamon despite beer being used (and not, in fact, being linked into cinnamon if you build with `-Wl,--as-needed`).

**Comment by leigh123linux** [comment]
@ItzSwirlz  pkexec command needs a polkit dialog.

**Comment by smcv** [comment]
It's not just pkexec: any system service that wants to do privileged actions in a controlled way using polkit (for example udisks, NetworkManager, systemd, firewalld, gdebi, flatpak --system) is going to need some sort of polkit agent to do the prompting/UI, either built-in to the desktop shell (as seen in GNOME Shell, GNOME Flashback, Phosh) or as a separate executable that is launched during desktop startup (as seen in Debian packages lxpolkit, lxqt-policykit, policykit-1-gnome, ukui-polkit, mate-polkit, polkit-kde-agent-1).

I think it certainly makes sense to include a polkit agent in "the Cinnamon desktop" (the overall system equivalent to the GNOME desktop), either built-in or as a dependency. Whether it's in cinnamon the program (the equivalent of gnome-shell) is a design question for its maintainers. If it's an external program, it will probably need to do careful things with X11 grabs to stop unrelated X11 apps from stealing focus and receiving your password, and it will probably be harder to present a UI where users can be confident that they're really interacting with "the desktop" and not some program that is imitating it.

The library dependency on polkit-agent-1 in Cinnamon does seem to be unused, at the moment. If Cinnamon gets a built-in polkit agent like the one in GNOME Shell, then the library dependency on polkit-agent-1 will be used again, to help to implement that; or if it uses an external polkit agent like policykit-1-gnome, it's the external polkit agent that will depend on the polkit-agent-1 library.

**Comment by Fantu** [comment]
From a fast look seems better readd integrated polkit agent rebasing it on the gnome-shell code.
Maintain it should not require lot of work, looking gnome-shell history the polkit agent files have few commits in a decade:
https://github.com/GNOME/gnome-shell/commits/master/src/shell-polkit-authentication-agent.c
https://github.com/GNOME/gnome-shell/commits/master/src/shell-polkit-authentication-agent.h

**Comment by ItzSwirlz** [comment]
On the Debian bug mail, mate-polkit was brought up. However, it may not be as stable. If mate-polkit is stable and healthy upstream, using it would be great FOSSwise

**Comment by smcv** [comment]
> the polkit agent files

The C code you linked is not all of it: that's just infrastructure. If the intention is to have a polkit agent built into Cinnamon (similar to how GNOME does it), you'll also need to bring back the UI. It's `js/ui/components/polkitAgent.js` in a recent-ish GNOME Shell, but it might have been in a different location in the version that Cinnamon originally forked from.

**Comment by ItzSwirlz** [comment]
Yeah, I listed that as part of the tasks. I also realized that the same commit got rid of something else at a similar time-the network agent. Noting this here in case a similar situation may arise

Polkit js files gone here: https://github.com/linuxmint/cinnamon/commit/72b406cc39e7571cd7d22ee132c2472617afddfe#diff-243027502bd084c987c5e544e6380d6e36c51517aea1b0747fe8a3444d8e75d7

**Comment by clefebvre** [comment]
> lxpolkit, lxqt-policykit, policykit-1-gnome, ukui-polkit, mate-polkit, polkit-kde-agent-1

seriously? We need a different agent in every single DE? Why?

**Comment by ItzSwirlz** [comment]
> seriously? We need a different agent in every single DE? Why?

Because UI really. Since this won't include a fallback mode, @mtwebster suggested maintaining policykit-1-gnome but I think if he wants to do that, it's better to switch to mate-polkit. MATE is pretty popular and I'm sure the polkit will stay around for a while.

**Comment by clefebvre** [comment]
Cinnamon and Xfce are not going to piggy back and use MATE components, there's no way you'll see inter-DE dependencies like this. If we need to maintain this it's either the original or as part of a cross DE projects such as Xapp.

Regarding UI, we've no need to theme the dialogs in Cinnamon, that's the only thing the Cinnamon was bringing really and that's why it was removed.

**Comment by ItzSwirlz** [comment]
> use MATE components,

Actually, mate-polkit doesn't need a mate component/tool for a dependency. It needs just gtk according to the imports in the code. But okay.



> it's either the original

I guess it is original then, unless the Xfce team is happy with an Xapp.

**Comment by leigh123linux** [comment]
Mate-polkit is even older the than polkit-gnome, fedora will continue to use polkit-gnome.

@smcv Why do you think it's better to use the code from gnome-shell?, it looks just as unmaintained as polkit-gnome.

https://github.com/GNOME/gnome-shell/commits/master/src/shell-polkit-authentication-agent.c

**Comment by ItzSwirlz** [comment]
Yeah, they are all not maintained as well. The only thing I see left to do then is just to be the maintainers/pickup for policykit-1-gnome. It's something all distros have maintained together.

**Comment by eli-schwartz** [comment]
Clearly, the code from gnome-shell is better maintained because it is stored inside of a git repo that isn't marked as archived. /s

Isn't that the traditional playbook of gnome-shell? Stop maintaining cleanly separated components in favor of something included directly into gnome-shell itself, which is *also* unmaintained but has better PR because shoving everything into one repo confers maintenance by association.

**Comment by clefebvre** [comment]
> Actually, mate-polkit doesn't need a mate component/tool for a dependency. It needs just gtk according to the imports in the code. But okay.

mate-polkit is a MATE component.

> I guess it is original then, unless the Xfce team is happy with an Xapp.

The XAPP project is cross-DE by definition, it's its mission, it's dedicated to supporting multiple DEs. There's no reason for DEs to  inter-depend on each other or to fork every little component out there into their own local version thanks to cross-DE projects like this one. GNOME is an exception to this because historically it developed everything before anyone else needed it.

> lxpolkit, lxqt-policykit, policykit-1-gnome, ukui-polkit, mate-polkit, polkit-kde-agent-1

This to me is crazy. Imagine adding cinnamon-polkit and xfce-polkit to the list, both doing the EXACT same as all the other ones. Unless we're looking for a specific look here (and even this could be done for each DE as part of a cross-DE effort) there's no need to duplicate this. You're talking about security and maintenance, are there any security issues with gnome-polkit? And if so, are they fixed in all these little forks?

Personally I'd recommend to continue to depend on gnome-polkit until there's an actual real problem with it. The day it gets forked it needs to be cross-DE.

**Comment by ItzSwirlz** [comment]
> Personally I'd recommend to continue to depend on gnome-polkit until there's an actual real problem with it

Will let you know if Cinnamon and Xfce get in trouble in Debian like they did with cjs later (probably won't for a while but I'll keep an eye on it)

Otherwise, yeah. A cross-DE is what we should probably use. Hopefully there can only be exceptions in terms of having their own polkits and instead for all, have one for Gtk branch DE's.

**Comment by leigh123linux** [comment]
> > Personally I'd recommend to continue to depend on gnome-polkit until there's an actual real problem with it
> 
> Will let you know if Cinnamon and Xfce get in trouble in Debian like they did with cjs later (probably won't for a while but I'll keep an eye on it)
> 

It is nothing like the mozjs52/cjs issue, debian will be able to support pokit-gnome for many years.

**Comment by smcv** [comment]
> Why do you think it's better to use the code from gnome-shell?, it looks just as unmaintained as polkit-gnome.

I am not *currently* aware of any security issues or other serious problems in PolicyKit-gnome: this isn't about whether serious problems are known *now*, but rather about what would happen if a serious problem is discovered in future. The thing I'm concerned about is that if someone discovers a serious problem next week or next month or next year, it will definitely not be fixed in PolicyKit-gnome (regardless of how serious it might be), because PolicyKit-gnome is archived and has no upstream maintainer. At that point, Cinnamon's only options would be to fork PolicyKit-gnome and fix the fork, or leave the problem unfixed in Cinnamon environments (or switch to some unrelated polkit agent, but long-term-stable distributions are unlikely to do that).

The polkit-related C code in gnome-shell is just glue to connect polkit to the Shell UI (the JavaScript part) and I'm not surprised it doesn't (need to) change often. The JavaScript code in the UI gets more commits than the C code.

> something included directly into gnome-shell itself, which is also unmaintained

That's not really true. Maybe it hasn't needed changes for a long time, but if a serious bug is discovered in GNOME Shell's polkit agent, it's the GNOME Shell maintainers who have taken responsibility for fixing it. At the moment, there is nobody who has the equivalent responsibility for the polkit agent used in Cinnamon, and that makes distributors uneasy.

One advantage of integrating the polkit agent into the "chrome" of the desktop compositor (as seen in GNOME Shell, and the equivalent hybrid JS/C agent that was removed from Cinnamon a while ago and reinstated in #10177) is that the compositor has total control over where focus and input go, so an integrated polkit agent is probably less likely to be affected by focus-stealing (your password going to an unrelated application that popped up a window at exactly the wrong time) or impersonation (an unrelated application making itself look exactly like the polkit agent so that you'll type your password into it); so if I was implementing a polkit agent myself, I'd probably be including it in the desktop shell too.

However, I'm not a Cinnamon developer (or a Cinnamon user, or a UI designer) and my design decisions are not the same as yours. If Cinnamon developers are confident that they can avoid that sort of issue for a polkit agent that is implemented as a separate process (the same overall design as PolicyKit-gnome, mate-polkit, etc.) then that's your call.

If Cinnamon developers have plans to move from X11 to Wayland in future, then that might have an effect on this design decision.

> debian will be able to support pokit-gnome for many years

To a point yes, but this assumes that there is someone in Debian who wants to take responsibility for maintaining polkit-gnome for the benefit of Cinnamon and maybe XFCE. The maintainers of Cinnamon in Debian would be an obvious candidate to do that, but the main maintainer of Cinnamon in Debian is no longer using Cinnamon, so other people who find Cinnamon important to them need to pick up the responsibility for Cinnamon's Debian packaging and maintenance (preferably people who can "agree to differ" while cooperating with other GTK-based desktop environments on shared components, rather than complaining that the other desktop environment is not behaving the way they would prefer). A software ecosystem that relies on people continuing to be responsible for components that they no longer use and no longer want to use is not sustainable.

I am not a Cinnamon user myself, but I opened this issue to try to help Cinnamon by alerting you to a potential future problem that you might not have previously been aware of. If this attempt to help is unwelcome, I can unsubscribe and go away, but I don't think it really benefits anyone for desktop environments to discourage other desktop environments' contributors from communicating.

**Comment by ItzSwirlz** [comment]
Thanks @smcv. I don't want to be the maintainer for policykit-1-gnome and I don't think anyone else wants to in any other distros. Eventually i will be the same goal as libdbus-glib-1-dev: get rid of it. I don't want to be hanging on to deprecated stuff, regardless of there are any security flaws in it or not. 

Clearly the best option to get the best for Cinnamon stability (as mentioned in the PR) and probably everything else is to fork and create a new cross-DE Xapp policy kit agent. And if we want to make progress before the policykit-1-gnome debian maintainers give up (which they already kind of have, [last update was in 2018](https://tracker.debian.org/pkg/policykit-1-gnome) and they are at Standards-Version 4.1.4, latest is 4.5.1), then I'd get going soon..

**Comment by SparkyBluefang** [comment]
While this discussion is specific to polkit-gnome, it is definitely not the only component Cinnamon depends on which has no maintained upstream (caribou, libtimezonemap). It seems like this is worth a broader discussion on how to deal with these class of dependencies. i.e.
* Keep using it and hope someone steps up to fix it when critical problems are discovered. Seems like this is where polkit-gnome has landed, and where caribou has been.
* Fork it and tailor it to cinnamon's needs. I'd argue this is where caribou belongs, as there's a lot of cruft in there cinnamon doesn't use. It looks like libtimezonemap might have been adopted by the Debian Cinnamon team, but that also seems like a weird status quo too.
* Replace it with a modern alternative if one is available.

---

Warning - ramblings ahead as a Gentoo package maintainer - by all means ignore if it's ignorant and/or not constructive :)

---

As an aside regarding re-using components from other DEs - when I was packaging libtimezonemap for Gentoo, I got the impression that it was a Unitiy component library, which would explain why it got abandoned around the same time Ubuntu dropped Unity.

This is where I feel a bit fuzzy around what constitutes a DE specific component vs a generic lib/util that happens to have a DE specific name. When you look at it, polkit-gnome, caribou, gnome-screenshot, and gsettings-desktop-schemas are from GNOME. libtimezonemap was(?) from Unity. So the argument to not use anything that originates from a different DE doesn't hold a whole lot of water for me. IMHO a more important determination is regarding code quality, active development, and whether or not there are DE exclusive dependencies/behaviors we wouldn't want also pulled in. For example, mate-polkit depends on mate-common as a non-starter :/

I'm also not sure I'd advocate for something like "polkit-xapp" as a replacement. I feel like the polkit client is something that should be dead simple, and having them try to do more DE integrated things is what led to all of the fragmentation. It seems like we should take a leaf from polkit-qt and propose "polkit-gtk" as a dead simple pure (py?)GTK implementation that all GTK DEs would use.

And, while I certainly wouldn't disagree that xapp is designed around being DE agnostic (though it does imply X?), from the standpoint of Gentoo, it is very much a Mint/Cinnamon specific paradigm. I think blueberry is the only non-Cinnamon package using it - and even then it's still from Mint. We've pretty much stuck with using the analogous GNOME apps.

**Comment by ItzSwirlz** [comment]
Xapp is (bear with me) a Mint thing and I am pretty confident the only reason it's in 90% of all popular distros is for cinnamon. Xapps shouldn't be the solution to all your problems, guys. While I am down for it, you need to think of what would be the best for everyone else _and yourself._

I don't care what the issue is but if you want to make life easier for everyone Cinnamon should bounce back to the MATE Polkit and only use it's parts or pick up maintainership. It's a popular DE not going around anywhere built for all types of Linux, not just Mint. But at this point the idea of what is needed is to have all these branches of GNOME 2(or 3, I suppose) to all use the same thing. In other words, don't make new stuff, maintain the old stuff and catch it up to current standards.

MATE - I am sure they are willing to drop polkit if policykit-1-gnome is remaintained as caribou is-an alternative to GNOME's fancy stuff. XFCE would catch on aswell.

This way Cinnamon keeps its stability and XFCE and MATE can have a bridge together with its own GNOME branch components. Hopefully we can get these repos out of the Archive section in the GNOME GitLab or host them here in Mint's GH.

**Comment by eli-schwartz** [comment]
Why does it matter if the remaintained polkit-gnome is called mate-polkit or polkit-gtk or polkit-xapp?

How does calling it "polkit-xapp" instantly make it less acceptable, due to the sheer name alone, to the MATE people than calling it polkit-gtk?

The xapp name is just a name denoting that the maintainers ate interested in allowing other DEs / applications to optionally choose to use the xapp software without locking themselves into DE specific behavior or dependencies. I'm mind boggled that anyone thinks it "has to" be mint specific.

**Comment by lestcape** [comment]
I'm in favour of a shared implementation always, but with an independent UI. I think this is what currently the GNOME implementation is. Cinnamon is a Clutter desktop, is not a Gtk desktop, but aparently the developers of Cinnamon are acting as it's Gtk, why?

In Cinnamon, the appindicator was droped in favor of a custtom implementation in XApp that is fully Gtk and when TingPing said: "was just disappointed that the mistakes of the past were repeated". I thought to myself that i was also disappointed to see that not one care about to support things in a way of a cross-toolkits, as not only a cross-platform is important today.

The desktop performance can not be selected first than congruence, because there are better ways to improve performance untouches that are waiting for an implementation (for example, separate the ui outside the compositor). In my vision, sacrifice the congruence for performance is not a good solution in the long run.

**Comment by collinss** [comment]
@eli-schwartz is correct here. The 'x' in xapps is meant to signify 'generic' or in other words distro- and de-agnostic. That is the whole intent of the project. See https://linuxmint-developer-guide.readthedocs.io/en/latest/xapps.html#xapps

@lestcape The Gnome implementation that we're talking about is exactly the opposite of shared and independent. It is entirely dependent on (and even integrated into) their own desktop environment.

As for Cinnamon being a clutter desktop, that wasn't really our design choice. We inherited it when we forked Gnome shell, and since then have tried to transition as much of the ui as possible from clutter and the main ui thread, into separate processes and using gtk for the ui. The biggest reasons we haven't done more in that regard is that we have limited resources, and a robust 3rd party applet/desklet/extension ecosystem that we don't want to abandon without really good reason.

**Comment by lestcape** [comment]
@collinss  I really don't know how good or bad is the GNOME implementation in this specific case, but usually they tend to block others. So, I believe you. Thanks for the clarification. A desk that is something but wants to be something else, is difficult for an outsider like me to understand. It is not clear to me why taking the work of clone so much code with Clutter and a monolithic architecture, if you want to have a Gtk desktop with a multi-process architecture. It would have been more effective to clone GNOME 2, who was much closer to that and was abandoned by GNOME in that time (don't explain me that really, at this point in the game it doesn't matter anymore).

Maybe Mint can create an implementation that will be really usable by everyone, not just a Gtk desktop. Whether you like Clutter or not, Clutter desktops are probably the most popular in Linux today. Cinnamon and PopOS are included in that list and i think cross-plataform means also cross-toolkit today because of that.

**Comment by chdslv** [comment]
> As for Cinnamon being a clutter desktop, that wasn't really our design choice. We inherited it when we forked Gnome shell, and since then have tried to transition as much of the ui as possible from clutter and the main ui thread, into separate processes and using gtk for the ui. 

Do you think Gnome would allow you to (or make it easy for you to) keep on forking the gnome shell?

**Comment by collinss** [comment]
Well, clutter has precisely one thing that is better than gtk for our purposes, and that is the ability to composite. That is the single greatest advantage that muffin has over window managers built on older technology. We certainly want a compositor for the window manager, but when it comes to drawing widgets on the screen, gtk is better without question. It is much more mature, stable, and feature rich.

The biggest issue though, is when too much of the ui is in the same process. If one component has a serious bug which causes a crash, it brings down the whole desktop. By separating as many of the ui components out of that process as possible, we reduce both the chances that something will bring down the main thread, and the number of components that are adversely affected by such a crash. This is the primary reason we don't want to put the polkit agent back into cinnamon.

**Comment by collinss** [comment]
@chdslv forking means we copied the gnome shell code as it was at the time and began developing it independently. From that moment on, any changes made to gnome shell have no effect on cinnamon unless we choose to copy them ourselves. Also, as I mentioned above, the x in xapps has nothing to do with the x window system.

**Comment by chdslv** [comment]
> The biggest issue though, is when too much of the ui is in the same process. If one component has a serious bug which causes a crash, it brings down the whole desktop. By separating as many of the ui components out of that process as possible, we reduce both the chances that something will bring down the main thread, and the number of components that are adversely affected by such a crash. This is the primary reason we don't want to put the polkit agent back into cinnamon.
![what is this](https://user-images.githubusercontent.com/28350701/125239906-a1846d80-e2e9-11eb-80cd-4ca0c618ad43.png)
What is this doing in LM then?

**Comment by clefebvre** [comment]
The way we fork things is pretty straight-forward:

- We only fork for a reason. If we're happy with something or if upstream is open to modifying it we don't fork it. 
- We fork with scope in mind. If working on a component can benefit 3 desktops instead of just 1 and we don't have a reason for that component to be DE-specific, we make it DE-agnostic and work on it under the XApp project, not the Cinnamon one.

When it comes to gnome-polkit, not only don't we have any plans for DE-specific changes, we don't have any plans for any changes at all. It works as it is. Sure, it's unmaintained, but until we actually do require changes that's not really an issue. The day it becomes an issue we'll either switch to a DE agnostic solution (like a polkit-gtk which doesn't exist right now) or we'll fork our own DE agnostic solution (xapp-polkit which would basically be the same as polkit-gtk).

In terms of scope our projects are pretty clear:

- Mint projects might work outside of Mint (mintinstall for instance) but they don't have to.
- Cinnamon projects might work outside of Cinnamon (nemo for instance) but they don't have to. They do need to work in other distros though.
- Xapp projects need to work in Cinnamon, MATE and Xfce (and other GTK DEs if there is an interest) and they need to work in other distros.

If you won't use an Xapp project just because it's made by Mint that's your problem. We've defined these scopes pretty clearly and we're very serious about cross-DE and cross-distro compatibility.

The presence of a required upstream dependency in Debian is a Debian specific problem. The fact that a component is absent in Debian doesn't mean it no longer exists or should be avoided. It just means Debian needs to do something about it.

That being said, the OP question is legitimate. If Debian, or any distro wants to clean things up and question a dependency we have, they're welcome to do so, as being done here.

I also find @SparkyBluefang's comment very relevant.

I'm going to take this opportunity to look into the very small gnome-polkit codebase and see if there's room for improvements. I'll add this to our roadmap. If there is, we'll fork it into an active project and then depending on the scope it'll either be a Cinnamon specific component or a DE agnostic Xapp component.

Regarding names, not all Xapp projects are called XSomething: blueberry, warpinator, bulky, sticky, hypnotix, slick-greeter, lightdm-settings, webapp-manager... if calling an Xapp polkit agent polkit-gtk rather than xapp-polkit makes people sleep better at night, it makes no difference to us.

We use prefixed and/or unique names to avoid name collisions with other projects. If polkit itself or GNOME (as GTK devs) was to make a polkit agent for GTK, they would naturally call it polkit-gtk, hence we're better off calling ours xapp-polkit or something like that. But again, we'd rather use upstream than fork our own, we'd rather not create meaningless DE-specific copies, and we'd rather use unique or prefixed names in Xapp subproject even if Xapp really is meant to be as cross-DE and cross-distro as GTK itself.

**Comment by clefebvre** [comment]
> @mtwebster @clefebvre why built-in polkit agent was removed?
> about forking another thing (PolicyKit-gnome, that is also not maintained upstream anymore) don't seems a good idea to me
> @leigh123linux @eli-schwartz other distros what think about this thing of PolicyKit-gnome?

Hi @Fantu, I'd have to check and see if the commit message actually recorded the original reason, but at least nowadays it would be removed for 3 separate reasons: Performance, maintenance and design/integration.

- Performance because Cinnamon is single-threaded in JS. It allowed for really easy extensions via monkey patching the code (much easier than the likes of libpeas plugins or sockets/plugs nightmares) but it also means we're a huge monolithic process where everything runs in the same place and affects each others.... from top priority sensitive WM operations all the way down to any slow I/O applet running in the corner of the screen. From a performance point of view, the more modular we are, the less runs in the WM thread, the better.

- Maintenance because JS means more bugs. Maybe not in general, some people love it. We don't. Depending on the use case we usually much prefer C or python. Modularity is also key for easy maintenance, taking something out of Cinnamon means we can work on it on its own and maintain it for all 3 DEs (or not maintain it at all if it exists upstream).

- Design/integration because Cinnamon isn't just a Clutter DE as @lestcape said. It's a Clutter/GTK DE. When you right-click your desktop you see a Nemo GTK context menu. It's a mixed environment with visual Clutter and GTK components. GNOME made a design choice to make it look more clutterish by implementing agents for notifications, session events and policykit. That's their choice and it gives GNOME a certain feel. We ship with a traditional paradigm and see the DE as a more minimal and more  modular entity. Panels and notifications traditionally look different (even in pure GTK DEs, you'll often see MATE dress itself with dark panels and notifications on top of a clear GTK theme for instance) but session events and polkit dialogs do not. Another reason not to make polkit clutterish looking is that it's tied to applications more so than to the DE. When you logout, you initiate the action from the menu. When you require polkit you might do so from the settings (GTK), or within an app (GTK). Last but not least clutter dialogs don't look as clean as GTK ones, though that's largely due to the quality of themes. You'll note that we moved to GTK also for melange. GTK is well maintained, mature, themed by many artists and its the toolkit of 90% of our environment, clutter/ST really is just a layer on top for the panel and a few other things.

**Comment by chdslv** [comment]
> The way we fork things is pretty straight-forward:
> 
> * We only fork for a reason. If we're happy with something or if upstream is open to modifying it we don't fork it.

There's a real chance Ubuntu might move to GTK 4 and fully to Wayland in the next LTS. Is LM Cinnamon ready to move to GTK4 and to Wayland in May-June 2022?

**Comment by lestcape** [comment]
> Design/integration because Cinnamon isn't just a Clutter DE as @lestcape said....

I don't know what aims to be the desktop, I only know in what direction most of people are trying to go. To me and others, merging uis should be a temporal things only, while the transition occurs as can be notice in the case of GNOME. Mate in the other hand is using the status icons with we all agree that is not a good feature. There are then some people that thinks that only XEmbed is the devil and others that thinks the devil is the fact of not respect the desktop ui, like can be me.

1. The desktop Menus in KDE that come from a different app that is outside are DBusMenu (not depended of the ui) always.
2. The same is happening in GNOME with Gtk4, as now all menus can be serialize, to not depend of the ui.
3. In MacOS the same is happening. The global menu is not displayed with the application ui, but with the desktop ui.
4. In Windows, well, who knows what happens there, I don't even remember the last time I worried about it. 

People in general are trying to always go to the solution that It allows to compose the pieces of the puzzle in the most coordinated and uniform way possible. It is not a matter of traditional or modern, but of coherence. But ofcourse not one is forced to go that way, we all know is the hard way after all.

Personally i think that move all to Gtk4 should be the best way to ensure consistency in Cinnamon, not matter the fact that we know that is not a compositor toolkit. Gtk4 have all the good things of Clutter except that point and for the other hand it have a lot of much more advantages. Just only i don' t see the point of wasted time in migrate Cinnamon and Mate at same time to Gtk4, as they have more less the same traditional layout. This should be relevant in that regards: https://gitlab.gnome.org/GNOME/gtk/-/issues/2132

**Comment by Fantu** [comment]
@clefebvre 
about performance and maintanance (mainly because of JS, I also had lost big amount of time to debug some issues with js when I used in website programming long time ago, is the language when I wasted more time in debug FWIR) you are right, the better option based on actually message in this topic seems to me keep maintain policykit-gnome (or fork it)
from a fast look is already indipendent from a single DE but depends on gtk, only small except is some gnome-common things (also deprecated that I suppose can be easy removed: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=830023
about issue (in policykit-gnome) actually the only relevant seems crash with wayland (that cinnamon don't support for now): https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=990259

@chdslv 
about wayland support probably is useful look this: https://github.com/linuxmint/muffin/pull/574#issuecomment-820224638 is about an advice for future wayland support but seems that require very long time to do it anyway (even if less that do it in actual "older fork") from a fast look (if I'm not wrong)

**Comment by mtwebster** [comment]
> There's a real chance Ubuntu might move to GTK 4 and fully to Wayland in the next LTS. Is LM and Cinnamon ready to move to GTK4 and to Wayland in May-June 2022?

I'm not sure you know what that means. That link has no bearing on any of this - they're discussing a technical issue with gtk4.  There are still apps that still use gtk2, or have only recently moved to use gtk3. I don't imagine either are going away by the next ubuntu LTS. They're libraries, they'll still be there regardless of whether or not gnome/ubuntu uses them - I think wayland is already their default, though I'm not sure. They have older hardware to consider as well. But again, I don't think x11 is going anywhere either - it doesn't matter whether they use it or not.

This is off-topic anyhow.

**Comment by chdslv** [comment]
@mtwebster 
Wayland might be off topic, but GTK 4 is not.

**Comment by clefebvre** [comment]
@chdslv the topic is Cinnamon's dependency on `policykit-1-gnome`.

**Comment by eli-schwartz** [comment]
> There's a real chance Ubuntu might move to GTK 4

That's not how computers work.

There is no software called Ubuntu with which to depend on gtk3 and "move" to gtk4. There is software called the gnome shell which does that, though.

Gtk3 is still under active development, because gtk4 is brand new and most software in the gtk ecosystem is not migrated. It could be years before we reach the tail end of applications taking that leap.

There is zero chance Ubuntu is going to decide to be the ONLY Linux distro where tons of software cannot be installed because "an -- until very recently -- actively developed and modern toolkit is irrationally on the banned list and may not be installed even as an optional universe package".

Ubuntu *may* be trying to ensure gtk4 and not gtk3 is shipped on their ISO and installed by default. But in that case, literally no one cares other than the Ubuntu release team plus blog authors scraping the bottom of the barrel for headlines.

Also what does any of this gtk3 fear mongering have to do with polkit? Is gtk3 somehow less secure than gtk4?

**Comment by clefebvre** [comment]
I had a quick look at the GNOME polkit agent. I can't think of any change I'd like to make to it, it works fine, looks fine and does exactly what we want. The source code is quite simple, an authenticator, a listener and a dialog. It's easy to maintain with no technical debt if it gets forked. It'd be easy to write from scratch if we wanted it in python also, it'd take maybe 3 or 4 days to be fully ready. Translations are dead (unmaintained just like the source code) but pretty complete. I really think the best course of action here is to continue to use upstream and not to fork unless we actually have a reason to.

Patches address Ubuntu specificities also (for instance with AccountServices) so the upstream project doesn't care about that. If we were to fork such a thing we'd start with the Ubuntu codebase (which includes the patches) and this wouldn't work as well as the original in other distributions until the code was changed to introduce conditions and adapt to distro specificities.

If a reason ever comes up, then before forking we'll look to see if a new cross-DE upstream project took over and if nobody did then I'd recommend we do this as part of Xapp. Until then policykit-1-gnome looks good to me.

**Comment by clefebvre** [comment]
I added a review of libtimezonemap and libcaribou in our roadmap. I was surprised not to see libtimezonemap in Debian and had to import it for LMDE, I didn't know it was tied to the Unity project though. Regarding Caribou there's a lot more to talk about... maintenance, size, the fact that we only really need the lib in Cinnamon, integration... it could be much smaller, much better and actively maintained. We'll look into this.

**Comment by Fantu** [comment]
@clefebvre 
https://tracker.debian.org/pkg/libtimezonemap was added in debian over a year ago
anyway a review if continue to use them in cinnamon (or search alternative or create one) I think is a good thing, thanks

**Comment by thomasphansen** [comment]
It might be worth mentioning that this thread became "heat" thanks to some reader of an Ubuntu blog, who found this issue (most likely without understanding it) and started using it as a flag against Linux Mint/Cinnamon. So, don't be surprised if you get more questions about completely unrelated things (like gtk4 and Wayland)... :/
Thanks for the nice work! :)

**Comment by ItzSwirlz** [comment]
> It might be worth mentioning that this thread became "heat" thanks to some reader of an Ubuntu blog, who found this issue (most likely without understanding it) and started using it as a flag against Linux Mint/Cinnamon. So, don't be surprised if you get more questions about completely unrelated things (like gtk4 and Wayland)... :/
> 
> Thanks for the nice work! :)

Send me the link to the blog post and I'll deal with it.

GTK4/Wayland can come later, as X11 and GTK3 aren't un maintained (yet). Otherwise, I would be glad to get started on a xapp-polkit or polkit-gtk or whatever. Just fork it somewhere and I'll contribute and package it for other distros, which will close said Debian issue. (It can be pushed to experimental regardless of bullseye freeze)

**Comment by eli-schwartz** [comment]
Huh, I knew it would be some blog clickbait...

**Comment by lestcape** [comment]
https://www.omgubuntu.co.uk/2021/07/linux-mint-20-2-released-this-is-whats-new

**Comment by clefebvre** [comment]
@smcv what agent does GNOME Flashback use in Debian? Afaik Shell's agent is part of Shell so it won't run in Flashback right? Do you ship Flashback with another agent? Do the polkits libs show a dialog and a basic UI agent in the absence of GNOME polkit? I'm asking because even though we depend on GNOME polkit, we only do so package-wise (in debian/control), and Debian has its own packaging, so it can perfectly remove that dependency and ship with the same solution as it uses for GNOME Flashback.

**Comment by chdslv** [comment]
> @mtwebster
> Wayland might be off topic, but GTK 4 is not. Ubuntu 22.04 LTS is relevant too, as the next LM LTS would be based on it.
> LM Cinnamon 20.2 has packages from Gnome 3.36. It is interesting, which Gnome shell version was used to fork Cinnamon 5?
> 
> Let me elaborate the question. Muffin in LM Cinnamon 20.2 has a high number 5.0.1+uma, but muffin is based on Mutter 3.2.1, which in turn last released in 2011. Cinnamon in Uma has a high number 5.0.4+uma. So, is the Cinnamon that was forked in 2011 is still the Cinnamon in Uma?

@ItzSwirlz This question was inspired by the blog post @thomasphansen is complaining about. After reading the arguments to and fro in that blog, I had a look at LM Cinnamon bit differently and the history. You are looking after Ubuntu-Cinnamon, so could be interested. I thought @clefebvre would reply, without closing the thread. I am really interested, maybe others too, which gnome-shell version was forked to create Cinnamon. It is hard to believe, every time, Gnome changed its shell, that was continuously forked. Won't be asking any more questions.

**Comment by collinss** [comment]
@chdslv it was only forked once. I don't know exactly which version, but it was an early one. Since then, cinnamon has diverged from gnome shell considerably. They are now very different projects. There have been a handful of changes that have been copied from Gnome shell, but they are rare.

**Comment by chdslv** [comment]
> @chdslv it was only forked once. I don't know exactly which version, but it was an early one. Since then, cinnamon has diverged from gnome shell considerably. 

Then, those people at that blog was right. Cinnamon was forked (copied, based) from a quite old, deprecated gnome shell. Of course, Cinnamon has diverged from Gnome considerably, just like gnome shell 3.38 is differ from 3.22, or Gnome 40 differ from 3.38. Nemo is old, forked from Nautilus 3.4, Mutter used in Muffin is deprecated. I do like old stuff. I use Openbox sometimes, but am worried it will stop working, if I keep on updating. 

Anyway as LM is based on Ubuntu LTS, LM has to depend on Ubuntu's packages, or hold all the packages in LM repos and block _any_ kind of upstream updating. Ubuntu would surely move away from deprecated packages, and stop holding them in their repos. So, it is a real worry, mine and others, whether LM is ready to move with Ubuntu changes, like GTK4, like Wayland etc. There's only Xubuntu out there lagging behind, so some of the packages will be kept yet, but not all. Anyway, thanks for replying @collinss

**Comment by thomasphansen** [comment]
> > It might be worth mentioning that this thread became "heat" thanks to some reader of an Ubuntu blog, who found this issue (most likely without understanding it) and started using it as a flag against Linux Mint/Cinnamon. So, don't be surprised if you get more questions about completely unrelated things (like gtk4 and Wayland)... :/
> > Thanks for the nice work! :)
> 
> Send me the link to the blog post and I'll deal with it.
> 
> GTK4/Wayland can come later, as X11 and GTK3 aren't un maintained (yet). Otherwise, I would be glad to get started on a xapp-polkit or polkit-gtk or whatever. Just fork it somewhere and I'll contribute and package it for other distros, which will close said Debian issue. (It can be pushed to experimental regardless of bullseye freeze)

The post mentioned by @lestcape is one of them, in the comments section. The same subject is raised again by the same user in another post, here:

https://www.omgubuntu.co.uk/2021/07/why-gnome-does-what-gnome-does-by-tobias-bernard#comment-5456455043

**Comment by eli-schwartz** [comment]
PLEASE OPEN A SEPARATE TICKET.

The topic of gtk4 is completely unrelated to "should cinnamon write a polkit agent in the main thread or keep relying on polkit-gnome or fork polkit-gnome".

**Comment by chdslv** [comment]
> PLEASE OPEN A SEPARATE TICKET.
> 
> The topic of gtk4 is completely unrelated to "should cinnamon write a polkit agent in the main thread or keep relying on polkit-gnome or fork polkit-gnome".

Isn't Policykit and GTK are interconnected somehow?

**Comment by smcv** [comment]
> Afaik Shell's agent is part of Shell so it won't run in Flashback right?

Correct. GNOME Shell's agent is essentially the same code that Cinnamon removed a few years ago, as previously mentioned on this issue (a Javascript/Clutter UI layer, currently a system-modal dialog similar to the ones for shutdown/reboot and password prompts, plus some C glue to interface that with libpolkit-agent).

> Do you ship Flashback with another agent?

I don't use or maintain Flashback myself, but from the packaging changelog it looks as though an agent (presumably GTK-based) was added to it in 3.18.0, so it does not need or use an agent from a separate package.

> Do the polkits libs show a dialog and a basic UI agent in the absence of GNOME polkit?

No, the polkit client library and daemon are entirely non-GUI. In particular, they do not depend on a specific GUI toolkit like GTK,  Qt or Clutter: the desktop environment is expected to provide a polkit agent, using a GUI toolkit and UI design that is appropriate for that desktop environment's requirements and conventions. The polkit agent could be part of the UI shell as in GNOME, or could be delegated to a separate program, whichever fits your design better.

Most polkit clients are system daemons like udisks2 and NetworkManager, for which showing a dialog isn't possible. A few polkit clients, notably `pkexec` and `flatpak --system`, will present a fallback TUI prompt if they happen to have been run from a terminal, but I wouldn't recommend relying on that.

> I'm asking because even though we depend on GNOME polkit, we only do so package-wise (in debian/control), and Debian has its own packaging, so it can perfectly remove that dependency and ship with the same solution as it uses for GNOME Flashback.

The closest equivalent of the solution we use for GNOME Flashback would be for Cinnamon to have its own built-in agent, similar to #10177. I suspect that's not the answer you hoped for.

> I had a quick look at the GNOME polkit agent. I can't think of any change I'd like to make to it, it works fine, looks fine and does exactly what we want. The source code is quite simple, an authenticator, a listener and a dialog. It's easy to maintain with no technical debt if it gets forked.

If that's the route you want to take, then I'd suggest that someone from Cinnamon should take responsibility for policykit-gnome (aka policykit-1-gnome) under a new name. GNOME developers are not going to maintain it for Cinnamon's benefit, but if there's nothing you'd want to change, then maintaining it under a new name should hopefully be low-effort.

**Comment by clefebvre** [comment]
> If a reason ever comes up, then before forking we'll look to see if a new cross-DE upstream project took over and if nobody did then I'd recommend we do this as part of Xapp. Until then policykit-1-gnome looks good to me.

This is the chosen resolution.

That said the dependency on GNOME polkit is only there in the Mint packaging, it doesn't exist in Cinnamon itself. In Debian, you're perfectly welcome to depend on lxde-polkit, mate-polkit or whatever alternative to gnome-polkit you want. Just depend on it in your own packaging and make sure to autostart it for Cinnamon sessions.

**Comment by kelebek333** [comment]
> The Debian maintainers of polkit noticed that the Debian package for Cinnamon 4.x depends on `policykit-1-gnome`, and the same seems to be true in Linux Mint's packaging of Cinnamon 5.x.

Did not the Debian maintainers of polkit notice other packages depends on "policykit-1-gnome" package? 

![policykit-1-gnome-rdepends](https://user-images.githubusercontent.com/1300133/126224321-565830aa-90e4-430b-983d-54d9509dace6.png)

**Comment by smcv** [comment]
> Did not the Debian maintainers of polkit notice other packages depends on "policykit-1-gnome" package?

Those dependencies are part of a list of alternatives that also accepts other polkit agents (apart from cinnamon-control-center, but whatever is done for cinnamon presumably applies equally to c-c-c).

**Comment by lestcape** [comment]
> To me and others, merging ui should be a temporal things only.

As i said, merging ui should be a temporal things only and now GNOME is leaving Gtk completely behind for a more pure Cutter desktop. That ofcourse it's really interesting as Cinnamon is supposedly in the opposite direction.

I think that if you want to continue to fork some GNOME packages, you will need to re-evaluate which direction you want to go. GNOME is getting rid of Gtk completely on the desktop, so it will most likely stop supporting embed any Gtk-related stuff into the desktop. This is surely going to impact anyone who instead wants to bring more Gtk stuff to the desktop, like can be Cinnamon.

The more important changes are in the pull requests:

1- They are moving system tray icons away from Gtk:
https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2590

2- Getting rid of Gtk completely in Mutter
https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2864

You are already alerted.

**Comment by Fantu** [comment]
hi, now in debian there are RC bugs against cinnamon and cinnamon-control-center for remove of policykit-1-gnome, are remained only them using it:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=990271
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1038630
@clefebvre @mtwebster have you decided what to do about cinnamon upstream and mint/LDME?
I ask because I would like to have as fewer differences as possible in the cinnamon packages in debian (with the upstream ones) if possible

**Comment by ItzSwirlz** [comment]
I think xdg-desktop-portal-xapp has polkit support in it (does it?)

**Comment by smcv** [comment]
> I think xdg-desktop-portal-xapp has polkit support in it (does it?)

I don't see any sign of it being a polkit agent. Polkit agents and portal backends are two separate components - in principle I think there would be nothing to prevent them from being hosted in the same executable, but I'm not aware of any desktop environment that actually does that.

Polkit agents and portal backends have some things in common: they should match the desktop environment's desired UI behaviour and appearance, and their purpose is to ask the user whether to allow something. They also have some important differences: polkit agents are about getting permission for a system service to do something with root privileges in response to an application request, whereas portal backends are about getting permission for a sandboxed app to do something (usually with only ordinary user privileges, and not system-wide privileges) outside its sandbox.

**Comment by lestcape** [comment]
For consistency with Mint's behavior, what I would expect is that instead of using a 3rd party application with a hard to digest 3rd party name (mate-polkit), Mint would fork policykit-1-gnome in some sort of XApp and then it can be used where it is needed, but with a generic name like that. This at least is supposed to be the objective of the XApps and I think this case falls into the same bag. It is clear that it is not a grace to have to maintain another package, but isn't it the same thing that has already happened in other desktops environments? 

At least that way Mint would have the plus of the generic name. Even if no one else decides to use it, the fact of think about sharing things with others, it's a plus to said thanks to Mint. I hope others see it that way, although they certainly can't be forced to use these "generic" packages and we cannot forced others to name his own packages in a generically way like Mint does either.

I need to said that this is not my preference, this just only what I hope will happen, from what I have seen that Mint has done in similar conditions. My preference is for a solution that uses Clutter as the GUI, as is the case in GNOME. It would certainly be nice if this solution ran in its own process or thread, but we already know that Clutter is not designed to be used as an app toolkit, but rather to run embedded within a desktop and also Clutter cannot run outside the main thread either.

I don't really know how much the polkit could spam the compositor thread, but I would expect that polkit actually only uses the CPU when is executing a direct user command and not in other contexts. That being the case, I don't see that including the polkit inside the shell is really an overhead or have negative effects on the performance of Cinnamon as a whole. I think this is what should be investigate and evaluate. Removing code that isn't actually running isn't going to make Cinnamon any faster.

**Comment by mtwebster** [comment]
I hate all this spam from force-pushing...

I've re-added polkit agent to Cinnamon, For now it's only going to be used for Wayland sessions, but it works in either session type.

**Comment by Fantu** [comment]
@mtwebster Big thanks for your works
why you wrote that for now it's going to be used only for Wayland sessions? From a fast look on https://github.com/linuxmint/cinnamon/commit/529d1412f116a5e382d6fa424d3aec3fd13a9299 policykit-1-gnome was removed from d/control

**Comment by mtwebster** [comment]
Oops, I'll add it back - at the moment our fallback mode (mate panel) still needs it. We might change our mind before Mint 22, but not sure.

**Comment by Fantu** [comment]
policykit-1-gnome will be removed before Debian 13 (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=990271), cinnamon is the only remained using it.
For cinnamon 6.2 is planned to fully remove it use? For now on master branch is only for Wayland session.

**Comment by SparkyBluefang** [comment]
I created #12272 while updating packages for Gentoo.
